# Laboratorio 2 - Parte 2

### Modelos no paramétricos

### 2019-II

#### Profesor: Julián D. Arias Londoño
#### julian.ariasl@udea.edu.co


## Guía del laboratorio

En esta archivo va a encontrar tanto celdas de código cómo celdas de texto con las instrucciones para desarrollar el laboratorio.

Lea atentamente las instrucciones entregadas en las celdas de texto correspondientes y proceda con la solución de las preguntas planteadas.

Nota: no olvide ir ejecutando las celdas de código de arriba hacia abajo para que no tenga errores de importación de librerías o por falta de definición de variables.

#### Primer integrante
Nombre: Kevin Martínez Gallego

#### Segundo integrante
Nombre: Andrés Mauricio Álvarez Ortiz

In [2]:
from __future__ import division
%matplotlib inline
import numpy as np
import numpy.matlib as matlib
import scipy as sc
import matplotlib.pyplot as plt

#Evitar algunas advertencias
import warnings
warnings.filterwarnings("always")

## Ejercicio 1: Contextualización del problema


Usaremos el dataset iris para el problema de clasificación, desde la librería sklearn. En el UCI Machine Learning Repository se encuentra más información en el siguiente [link](https://archive.ics.uci.edu/ml/datasets/iris) .

In [4]:
#cargamos la bd iris desde el dataset de sklearn
from sklearn import datasets

iris = datasets.load_iris()

X, Y = iris.data, iris.target

In [5]:
print(X.shape)
print(Y.shape)
print(Y)
print(np.unique(Y, return_counts = True))

(150, 4)
(150,)
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
(array([0, 1, 2]), array([50, 50, 50], dtype=int64))


Responda las siguientes preguntas: 

1.1 ¿Cu&aacute;ntas clases tiene el problema?:


3 clases

1.2 ¿Cu&aacute;ntas caracter&iacute;sticas tiene el problema?:

4 características

1.3 ¿Cu&aacute;ntas muestras tiene el problema?:

150 muestras

1.4 ¿Cu&aacute;ntas muestras por cada clase hay en la base de datos?:

Hay 50 muestras por cada clase

## Ejercicio 2: Completar código KNN

Analice los siguientes métodos de la teoría vista para los modelos de *K-vecinos más cercanos (KNN)* y complete el código del método KNN.

<b>Nota</b>: Para el cáculo de la distancia entre vectores tienen dos opciones, usar la función la distancia euclidiana `scipy.spatial.distance.euclidean`([Ejemplo](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.euclidean.html)) o usar la función `numpy.linalg.norm`([Ejemplo](https://docs.scipy.org/doc/numpy/reference/generated/numpy.linalg.norm.html)). Revisen la documentación para comprender su uso. También serán de utilidad las funciones `numpy.sort` y `numpy.argsort`.


<b>Nota</b>: Para el cáculo de la distancia entre vectores tienen dos opciones, usar la función scipy.spatial.distance.euclidean o usar la función numpy.linalg.norm. <b>Revisen la documentación </b> para comprender su uso. También serán de utilidad las funciones numpy.sort y numpy.argsort.

In [6]:
from scipy.stats import mode
from scipy.spatial import distance
from scipy import stats
import operator

#Para calcular el error en los problemas de clasificación
def ErrorClas(Y_lest, Y):
    error = 1 - np.sum(Y_lest == Y)/len(Y)
    
    return error


def KNN(X_train, Y_train, X_val, k, tipo):
    
    #X_train es la matriz con las muestras de entrenamiento
    #Y_train es un vector con los valores de salida pra cada una de las muestras de entrenamiento
    #X_val es la matriz con las muestras de validación
    #tipo es una bandera que indica si el problema es de regresión o de clasificación.
    tipo = tipo
    #Parámetro k que equivale al número de vecinos a tener en cuenta para resolver el problema de
    #predicción de la variable de salida
    k = k  
        
    Ntest = len(X_val)
    Y_test = np.zeros(Ntest)
    Y_train_kVecinos = np.zeros(k)

    Ntrain = len(X_train)
    distancias = np.zeros(Ntrain)
    for i in range(Ntest):
        muestra = X_val[i,:]
        for j in range(Ntrain):
            distancias[j] = distance.euclidean(muestra, X_train[j])
        distanciasSortIndex = np.argsort(distancias)
        for j in range(k):
            Y_train_kVecinos[j] = Y_train[distanciasSortIndex[j]]
        if(tipo == 0): # clasificación
            Y_test[i] = stats.mode(Y_train_kVecinos)[0]
        elif(tipo == 1): # regresión
            Y_test[i] = np.mean(Y_train_kVecinos)
    
    #Debe retornar un vector que contenga las predicciones para cada una de las muestras en X_val, en el mismo orden.  
    return Y_test
  

## Ejercicio 3

Una vez haya completado el codigo del método de KNN, ejecute varias veces el proceso de entrenamiento y evaluación cambiando el parametro $k$, el cual es el numero de vecinos, y complete la siguiente tabla con los valores del error de clasificación obtenidos:

In [8]:
from numpy import random
from sklearn.model_selection import train_test_split
import math

def executeKNN(k, tipo):
    N = np.size(X,0)

    #Validamos el modelo
    random.seed(1)
    Error = np.zeros(4)
    for j in range(4):

        Xtrain, Xtest, Ytrain, Ytest = train_test_split(X, Y, test_size=0.24)

        #Normalizamos los datos
        media = np.mean(Xtrain)
        desvia = np.std(Xtrain)
        Xtrain = sc.stats.stats.zscore(Xtrain)
        Xtest = (Xtest - np.matlib.repmat(media, Xtest.shape[0], 1))/np.matlib.repmat(desvia, Xtest.shape[0], 1)

        k = k
        tipo = tipo

        Yest = KNN(Xtrain, Ytrain, Xtest, k, tipo)#Complete con el llamado apropiado de la función que Uds diseñaron

        #Evaluamos las predicciones del modelo con los datos de test
        Error[j] = ErrorClas(Yest,Ytest)
        
    error = np.mean(Error)
    std = np.std(Error)
    print('\nError durante la prueba = ' + str(error) + '+-' + str(std))
    return(error, std)

Responda:

3.1 ¿Qué metodología de validación se usa en el experimento?


Se utiliza validación Booststrap.

3.2 ¿Cuántas muestras se usan para el entrenamiento?


Para entrenamiento se utiliza el 76% de las muestras, lo que equivale a 114 muestras.

3.3 ¿Cuántas muestras se usan para la validación?

Para validación se utiliza el 24% de las muestras, lo que equivale a 36 muestras.

#### Tabla de resultados

In [10]:
import pandas as pd
import qgrid
randn = np.random.randn
df_types = pd.DataFrame({
    'Numero de vecinos' : pd.Series(['1', '2', '3', '4', '5', '6', '7'])})
df_types["Error_Prueba"] = ""
df_types["Desviación estándar del error"] = ""
df_types.set_index(['Numero de vecinos'], inplace=True)
#df_types["Error_Prueba"][1] = "0.2697"
#df_types["Desviación estándar del error"][1] = "0.088"
#df_types.sort_index(inplace=True)

tipo = 0 # clasificación
valoresK = np.array([1,2,3,4,5,6,7])
for i in range(np.size(valoresK)):
    error, std = executeKNN(valoresK[i], tipo)
    df_types["Error_Prueba"][i] = error
    df_types["Desviación estándar del error"][i] = std

qgrid_widget = qgrid.show_grid(df_types, show_toolbar=False)
qgrid_widget


Error durante la prueba = 0.20833333333333337+-0.11197580205970208

Error durante la prueba = 0.2777777777777778+-0.08561686115234694

Error durante la prueba = 0.27083333333333337+-0.09081039465709739

Error durante la prueba = 0.2916666666666667+-0.07216878364870326

Error durante la prueba = 0.2777777777777778+-0.08561686115234694

Error durante la prueba = 0.29166666666666663+-0.08893228107545624

Error durante la prueba = 0.2847222222222222+-0.08187379251771945


QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': True, 'defau…

Ejecute la siguiente instrucción para dejar guardados en el notebook los resultados de las pruebas.

In [11]:
qgrid_widget.get_changed_df()

Error_Prueba Desviación estándar del error
Numero de vecinos                                           
1                     0.208333                      0.111976
2                     0.277778                     0.0856169
3                     0.270833                     0.0908104
4                     0.291667                     0.0721688
5                     0.277778                     0.0856169
6                     0.291667                     0.0889323
7                     0.284722                     0.0818738

## Ejercicio 4

A continuación debe completar el código del método parzen para resolver problemas de clasificación con el modelo ventana de Parzen.

In [15]:
from scipy.spatial import distance

#Calcula el kernel gaussiano de x
def kernel_gaussiano(x):
    return np.exp((-0.5)*x**2)

def ParzenWindow(x,Data,h):
    h = h
    Ns = Data.shape[0]
    suma = 0
    for k in range(Ns):
        u = sc.spatial.distance.euclidean(x,Data[k,:])
        suma += kernel_gaussiano(u/h)
    return suma/Ns

def ParzenClass(X_train, Y_train, X_val, ancho_h):
    
    h = ancho_h
    Nval = len(X_val)
    Ntrain = len(X_train)
    Y_val = np.zeros(Nval)
    
    clases = np.unique(Y_train)
    Nclases = len(clases)
    
    clases_separadas = []
    estimados = np.zeros(Nclases)
    
    for i in range(Nclases):
        index = Y_train == clases[i]
        clases_separadas.append(X_train[index, :]) # Conjuntos de entrenamiento separado por clases
        # obtenemos todas las muestras de entrenamiento pertenecientes a cada clase
    
    for i in range(Nval):
        muestra = X_val[i,:]
        for j in range(Nclases):
            estimados[j] = ParzenWindow(muestra,clases_separadas[j],h)
        Y_val[i] = np.argmax(estimados) # argmax retorna el índice donde se encuentre el mayor
    
    #Debe retornar un vector que contenga las predicciones para cada una de las muestras en X_val, en el mismo orden.  
    return Y_val

## Ejercicio 5

Ahora debe resolver el mismo problema de clasificación con el método ventana de Parzen. Complete el código ejecute las pruebas para diferentes valores de $h$ y llene la siguiente tabla:

In [22]:
from numpy import random
from sklearn.model_selection import StratifiedKFold
import math

def executePARZEN(ancho_h):
    #Validamos el modelo
    random.seed(1)
    Error = np.zeros(4)
    skf = StratifiedKFold(n_splits=4)
    j = 0
    for train, test in skf.split(X, Y):
        Xtrain = X[train,:]
        Ytrain = Y[train]
        Xtest = X[test,:]
        Ytest = Y[test]

        #Normalizamos los datos
        media = np.mean(Xtrain)
        desvia = np.std(Xtrain)
        Xtrain = sc.stats.stats.zscore(Xtrain)
        Xtest = (Xtest - np.matlib.repmat(media, Xtest.shape[0], 1))/np.matlib.repmat(desvia, Xtest.shape[0], 1)

        ancho_h = ancho_h

        Yest = ParzenClass(Xtrain, Ytrain, Xtest, ancho_h)#Complete con el llamado apropiado de la función que Uds diseñaron

        #Evaluamos las predicciones del modelo con los datos de test
        Error[j] = ErrorClas(Yest,Ytest)
        j += 1  
        
    error = np.mean(Error)
    std = np.std(Error)
    print('\nError durante la prueba = ' + str(error) + '+-' + str(std))
    return(error, std)
        


Error durante la prueba = 0.2216880341880342+-0.04332854434165311


Responda:
    
5.1 ¿Qué metodología de validación se usó en la simulación?:



Se utiliza la metodología de validación cruzada, empleando 4 subconjuntos.

In [23]:
print(Xtrain.shape)
print(Ytrain.shape)
print(Xtest.shape)
print(Ytest.shape)

(114, 4)
(114,)
(36, 4)
(36,)


5.2 ¿Cuántas muestras se usan para el entrenamiento?:



114 muestras.

5.3 ¿Cuántas muestras se usan para la validación?:


36 muestras.

#### Tabla de resultados

In [18]:
import pandas as pd
import qgrid
randn = np.random.randn
df_types2 = pd.DataFrame({
    'Tamano de ventana' : pd.Series(['0.05', '0.1', '0.5', '1', '2', '5', '10'])})
df_types2["Error_Prueba"] = ""
df_types2["Desviación estándar del error"] = ""
df_types2.set_index(['Tamano de ventana'], inplace=True)
#df_types2["Error_Prueba"][1] = "0.222"
#df_types2["Desviación estándar del error"][1] = "0.043"
#df_types.sort_index(inplace=True)

valoresH = np.array([0.05,0.1,0.5,1,2,5,10])
for i in range(np.size(valoresH)):
    error, std = executePARZEN(valoresH[i])
    df_types2["Error_Prueba"][i] = error
    df_types2["Desviación estándar del error"][i] = std

qgrid_widget = qgrid.show_grid(df_types2, show_toolbar=False)
qgrid_widget


Error durante la prueba = 0.2216880341880342+-0.04332854434165311

Error durante la prueba = 0.2216880341880342+-0.04332854434165311

Error durante la prueba = 0.281517094017094+-0.04021339704891218

Error durante la prueba = 0.280982905982906+-0.03145818774430656

Error durante la prueba = 0.24198717948717952+-0.058466142137675314

Error durante la prueba = 0.2216880341880342+-0.04696907336459245

Error durante la prueba = 0.2216880341880342+-0.04696907336459245


QgridWidget(grid_options={'fullWidthRows': True, 'syncColumnCellResize': True, 'forceFitColumns': True, 'defau…

Ejecute la siguiente instrucción para dejar guardados en el notebook los resultados de las pruebas.

In [19]:
qgrid_widget.get_changed_df()

Error_Prueba Desviación estándar del error
Tamano de ventana                                           
0.05                  0.221688                     0.0433285
0.1                   0.221688                     0.0433285
0.5                   0.281517                     0.0402134
1                     0.280983                     0.0314582
2                     0.241987                     0.0584661
5                     0.221688                     0.0469691
10                    0.221688                     0.0469691